### PR TITLE
batched undelegations - undelegation ICA submission

### PR DESCRIPTION
**Pull request by sampocs**
## Context
This PR is responsible for the changes in the part of the codebase responsible for the undelegate ICA submission. 

The high level approach is that:
* The ICA's are now submitted in batches
* The callback data now stores the st token amounts for each validator
* In the callback (separate PR), the st tokens will be burned for the respective validators as each ICA finishes
* If a batch fails, the failed amount will be tacked onto next epoch's unbondings

**NOTE TO REVIEWER** I would recommend reviewing this PR commit-by-commit!

## Changelog
* Updated `GetTotalUnbondAmount` to include `UNBONDING_RETRY_QUEUE` records in the summation
* Updated `RefreshUnbondingNativeTokenAmounts` to only "refresh" the native value of `UNBONDING_QUEUE` records (`RETRY_QUEUE` records should not be touched)
* Updated `RefreshUnbondingNativeTokenAmounts` to return the updated version of the HZU mapping that was passed into the function
   * We previously would call (1) get record mapping (2) refresh records (3) get record mapping again
   * But I think it's safer to just return the updated mapping directly from the refresh function
* Updated `GetUnbondingICAMessages` to calculate the st token amount for each validator (using the implied redemption rate from the EUR). The last validator gets the remainder
* Stored the st token amount in the callback data
* Updated unit tests and misc build errors due to proto changes

### CODE REVIEW

**Review by ethan-stride — APPROVED** [review]
LGTM so far but still need to see the callback updating logic.
Thanks for splitting it into these readable commits, definitely helped to walk them one by one.

From looking ahead I now see some of my questions are answered about atomicity -- not only is a whole hostZoneRecord unbonding going to be atomic but because of the applyIfNoError wrapper the entire set of ICAs will either all succeed or all fail which simplifies things (at the increased risk of a problem one holding everyone up).
Most of my comments are just to confirm my understanding, I don't see anything logically wrong but I also only have about 80% of the full picture so far, I think you are still working on some of the related PRs in drafts which have that last 20% of the picture.

Definitely have Riley/Others also review this and I might give it one more look when all the PRs are done but so far amazing and fast work here.  This should make a lot of things more understandable with the state tracking new variables on the hostZone level records as well as of course the batching specific new abilities

File: x/records/types/records.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ import sdkmath "cosmossdk.io/math"
 // Helper function to evaluate if a host zone unbonding record still needs to be initiated
 func (r HostZoneUnbonding) ShouldInitiateUnbonding() bool {
 	notYetUnbonding := r.Status == HostZoneUnbonding_UNBONDING_QUEUE
+	hasFailedUnbonding := r.Status == HostZoneUnbonding_UNBONDING_IN_PROGRESS
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/records/types/records.go]: I think you may be looking at a stale commit? there's two functions now - one for the regular case and one for the failure/retry case with the new status

File: x/records/types/records.go [hunk 2]
```diff
@@ -5,6 +5,7 @@ import sdkmath "cosmossdk.io/math"
...
 	hasAtLeastOneRecord := r.NativeTokenAmount.GT(sdkmath.ZeroInt())
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/records/types/records.go]: The rounding error is prevalent in any of these functions since all values are denominated in the value as they appear on stride

File: x/stakeibc/keeper/unbonding.go [hunk 3]
```diff
@@ -65,14 +66,18 @@ func (k Keeper) GetQueuedHostZoneUnbondingRecords(
...
+// that should be unbonded, and the total number of stTokens that should be burned
+func (k Keeper) GetTotalUnbondAmount(
+	hostZoneUnbondingRecords map[uint64]recordstypes.HostZoneUnbonding,
+) (totalNativeAmount sdkmath.Int, totalStAmount sdkmath.Int) {
+	totalNativeAmount = sdk.ZeroInt()
+	totalStAmount = sdk.ZeroInt()
 	for _, hostZoneRecord := range hostZoneUnbondingRecords {
-		totalUnbonded = totalUnbonded.Add(hostZoneRecord.NativeTokenAmount)
+		totalNativeAmount = totalNativeAmount.Add(hostZoneRecord.NativeTokensToUnbond)
+		totalStAmount = totalStAmount.Add(hostZoneRecord.StTokensToBurn)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:78]: yeah we use the RR at the time the unbonding is submitted and "refresh" all the records to use that more updated value

File: x/stakeibc/keeper/unbonding.go [hunk 4]
```diff
@@ -123,19 +128,35 @@ func (k Keeper) RefreshHostZoneUnbondingNativeTokenAmount(
 
 	// Then set the total on the host zone unbonding record
 	hostZoneUnbondingRecord.NativeTokenAmount = totalNativeAmount
+	hostZoneUnbondingRecord.NativeTokensToUnbond = totalNativeAmount
+	hostZoneUnbondingRecord.StTokensToBurn = hostZoneUnbondingRecord.StTokenAmount
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:132]: that's right - it should never get touched after it's initially set

File: x/stakeibc/keeper/unbonding.go [hunk 5]
```diff
@@ -225,12 +246,17 @@ func SortUnbondingCapacityByPriority(validatorUnbondCapacity []ValidatorUnbondCa
...
-) (msgs []proto.Message, unbondings []*types.SplitDelegation, err error) {
+) (msgs []proto.Message, unbondings []*types.SplitUndelegation, err error) {
+	// Calculate the implied redmeption rate from the totals
+	impliedRedemptionRate := sdk.NewDecFromInt(totalNativeAmount).Quo(sdk.NewDecFromInt(totalStAmount))
+
 	// Loop through each validator and unbond as much as possible
-	remainingUnbondAmount := totalUnbondAmount
+	remainingUnbondAmount := totalNativeAmount
+	remainingStAmount := totalStAmount
 	for _, validatorCapacity := range prioritizedUnbondCapacity {
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:259]: from ethan on slack: 
> Oh I thought we only did it downstream.  If we do it upstream then yeah it is resolved

Resolving

File: x/records/types/records.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ethan-stride) [reply x/records/types/records.go]: So how will we handle that potential issue?  Comparison directly to zero seems like we will run directly into the rounding issue sooner or later.  In this case, the hostZoneUnbonding record could basically be empty but the rounding error could make us think there is still some unbonding which need to happen?

File: x/stakeibc/keeper/unbonding.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ethan-stride) [reply x/stakeibc/keeper/unbonding.go:78]: See the big open question on PR 1199 which is directly related to this.

File: x/stakeibc/keeper/unbonding.go [hunk 6]
```diff
@@ -247,147 +273,43 @@ func (k Keeper) GetUnbondingICAMessages(
...
-	if len(unbondings) > batchSize {
-		unbondings, err = k.ConsolidateUnbondingMessages(totalUnbondAmount, unbondings, prioritizedUnbondCapacity, batchSize)
-		if err != nil {
-			return msgs, unbondings, errorsmod.Wrapf(err, "unable to consolidate unbonding messages")
+		// Calculate the associated stToken amount
+		// There may be some precision error so the last validator should
+		// get all the remaining stTokens
+		var stAmount sdkmath.Int
+		if remainingUnbondAmount.IsZero() {
+			stAmount = remainingStAmount
```
Comment (ethan-stride) [inline x/stakeibc/keeper/unbonding.go:280]: This guarantees that remainingStAmount will hit zero at the same time that remainingUnbondAmount reaches zero so there is no need for a break on the remainingStAmount being zero or even a sanity check for it.

File: x/stakeibc/keeper/unbonding.go [hunk 7]
```diff
@@ -225,12 +246,17 @@ func SortUnbondingCapacityByPriority(validatorUnbondCapacity []ValidatorUnbondCa
...
 	hostZone types.HostZone,
-	totalUnbondAmount sdkmath.Int,
+	totalNativeAmount sdkmath.Int,
+	totalStAmount sdkmath.Int,
 	prioritizedUnbondCapacity []ValidatorUnbondCapacity,
 	batchSize int,
-) (msgs []proto.Message, unbondings []*types.SplitDelegation, err error) {
+) (msgs []proto.Message, unbondings []*types.SplitUndelegation, err error) {
+	// Calculate the implied redmeption rate from the totals
+	impliedRedemptionRate := sdk.NewDecFromInt(totalNativeAmount).Quo(sdk.NewDecFromInt(totalStAmount))
```
[the remark this answers is not in the document]
Comment (ethan-stride) [reply x/stakeibc/keeper/unbonding.go:254]: I didn't follow that last sentence completely but it feels important.  Can you rephrase it or go into more detail?  It would seem that the stToken and nativeAmounts must always hit zero at the same time no?

File: x/records/types/records.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply x/records/types/records.go]: so sorry I meant **the rounding error is **not** prevalent

[the remark this answers is not in the document]
Comment (sampocs) [reply x/records/types/records.go]: you can see my response in the other PR for an explanation of the rounding issue

File: x/stakeibc/keeper/unbonding.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:254]: I responded to this in the other PR as well, but here's an example:

```
EUR 1 - RR: 1.0, Native: 1000, StTokens: 1000
EUR 2 - RR: 1.5, Native: 1500, StTokens: 1000
EUR 3 - RR: 2.0, Native: 2000, StTokens: 1000
```

When we submit the unbonding, we submit it for the total which is 4500 native, and 3000 sttokens.
This has an implied rate of 4500/3000 = 1.5

However, let's say the first batch only unbonds 2500 tokens.
Since we calculate the number of stTokens using the implied rate of 1.5, this means we store 1666 sttokens on that batch

So when we decrement those amounts from both records we end up with:
```
EUR 1 - Native: 0, StTokens: 0            // subtracts 1000n, 1000st and "zero's" them both out 
EUR 2 - Native: 0, StTokens: 334          // subtracts the remaining 1500n, and 666st - only the native tokens are zero'd out
EUR 3 - Native: 2000, StTokens: 1000
```

So my point was that they wont zero out at the same time, but once all the other batches completes, then the last record will zero out together

[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:254]: So I think the other way to solve this is to just compute the number of stTokens in the callback using the ratio in the EUR. The only issue is then you get some precision error on the last validator so you probably need a clause that checks if the stToken amount is smaller than some epsilon, then round it down to 0

**Review by riley-stride — COMMENTED** [review]
Looking very strong. A few open questions but spec + excalidraw + commit by commit review made this digestible (relative to the complexity).

File: x/stakeibc/keeper/unbonding.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:254]: hmm, yeah actually maybe it does work! cause you're right in that it would just use the totals next epoch, so the zero doesn't really make a difference

File: x/stakeibc/keeper/unbonding.go [hunk 8]
```diff
@@ -468,9 +386,9 @@ func (k Keeper) UnbondFromHostZone(ctx sdk.Context, hostZone types.HostZone) err
 	undelegateBatchSize := int(hostZone.MaxMessagesPerIcaTx)
 	msgs, unbondings, err := k.GetUnbondingICAMessages(
 		hostZone,
-		totalUnbondAmount,
+		totalNativeUnbondAmount,
+		totalStBurnAmount,
 		prioritizedUnbondCapacity,
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:391]: yeah good eye! I'm going to change the sorting in a different PR though

File: x/stakeibc/keeper/unbonding.go [hunk 9]
```diff
@@ -57,22 +58,26 @@ func (k Keeper) GetQueuedHostZoneUnbondingRecords(
...
 	return epochNumbers, epochToHostZoneUnbondingMap
 }
 
-// Gets the total unbonded amount for a host zone by looping through the epoch unbonding records
-// Also returns the epoch unbonding record ids
-func (k Keeper) GetTotalUnbondAmount(ctx sdk.Context, hostZoneUnbondingRecords map[uint64]recordstypes.HostZoneUnbonding) (totalUnbonded sdkmath.Int) {
-	totalUnbonded = sdk.ZeroInt()
+// Loops through each epoch unbonding record and returns the total number of native tokens
+// that should be unbonded, and the total number of stTokens that should be burned
+func (k Keeper) GetTotalUnbondAmount(
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:71]: sorry the excalidraw I think had a typo - we want to sum UNBONDING_QUEUE and UNBONDING_RETRY_QUEUE. We do **not** want to sum UNBONDING_IN_PROGRESS

It's decided [here](https://github.com/Stride-Labs/stride/pull/1196/files#diff-cdec6e1b56af7aab76f27f380b93d2baf9dcd031042a5770361c63b8dde3d5dfR61) when we get the queued records (and then these records are passed into this function)

File: x/stakeibc/keeper/unbonding.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:254]: This morning we discussed a different approach to re-compute the number of stTokens in the callback based on the RR from the individual record. Assuming both work, do you have a preference?

Approach #1:
* only does the stToken calculation once per epoch
* handles precision error gracefully
* but, has this weird edge case above during retries where the amounts don't match

Approach #2:
* does stToken calculation each callback
* handling precision error is more hand wavy
* but, amounts always match the record's RR

cc @ethan-stride

[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/unbonding.go:254]: I don't have a strong preference but lean towards #2 to prioritize having amounts that are interpretable and diagnosable. As long as we are confident the precision errors will not compound and won't cause errors that halt the system. Wby?

File: x/stakeibc/keeper/unbonding.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/unbonding.go:71]: Sounds good. Resolving.

File: x/stakeibc/keeper/unbonding.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/unbonding.go:254]: Might be good to get additional input from @shellvish @asalzmann on this one.

**Review by shellvish — APPROVED** [review]
This looks fantastic, thanks so much for doing this Sam!

File: x/stakeibc/keeper/unbonding.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/unbonding.go:254]: discussed offline - moving forward with approach #2 from above for now and will revisit refreshing the RR again during retries later
